### PR TITLE
Introduce `SendAsyncSafe` trait

### PR DESCRIPTION
**Pull request by luckysori**
This was mentioned in https://github.com/itchysats/itchysats/issues/553#issuecomment-983191673.

### CODE REVIEW

File: daemon/src/send_async_safe.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+#[async_trait]
+impl<A, M> SendAsyncSafe<M> for xtra::Address<A>
+where
+    A: xtra::Handler<M>,
+    M: xtra::Message<Result = ()>,
+{
+    async fn send_async_safe(&self, msg: M) -> Result<(), xtra::Disconnected> {
+        #[allow(clippy::disallowed_method)]
+        self.do_send_async(msg).await
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs:26]: Do a `tokio::spawn` here and use `.send` and log the error if it happens.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]

File: daemon/src/send_async_safe.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/send_async_safe.rs:26]: Sorry, for this one there is no error but we can do that for the TODO above!

Comment (luckysori) [reply daemon/src/send_async_safe.rs:26]: I don't understand these two comments, could you rephrase?

Comment (thomaseizinger) [reply daemon/src/send_async_safe.rs:26]: Ignore this, I'll post again!

File: daemon/src/send_async_safe.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+use async_trait::async_trait;
+
+#[async_trait]
+pub trait SendAsyncSafe<M>
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs]: Can we extend this with a type parameter `R`?

File: daemon/src/send_async_safe.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+where
+    M: xtra::Message<Result = ()>,
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs]: And use `R` here?

File: daemon/src/send_async_safe.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    ///
+    /// As soon as this method returns, we know if the receiving actor
+    /// is alive. If they are, the message we are sending will
+    /// eventually be handled by them, but we don't wait for them to
+    /// do so.
+    async fn send_async_safe(&self, msg: M) -> Result<(), xtra::Disconnected>;
+}
+
+#[async_trait]
+impl<A, M> SendAsyncSafe<M> for xtra::Address<A>
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs]: And implement this twice:

```rust
impl<A, M> SendAsyncSafe<M, ()> for xtra::Address<A> {}
impl<A, M, E: Error> SendAsyncSafe<M, Result<(), E>> for xtra::Address<A> {}
```
And in the 2nd impl we can do:

```rust
let fut = self.clone().send(msg);

tokio::spawn(async move {
	if let Err(e) = fut.await {
		tracing::warn!("Asynchronous message invocation failed: {:#}", e);
	}
})
```

Comment (luckysori) [reply daemon/src/send_async_safe.rs]: Two things:

- I cannot implement the same trait for the same generic type twice (even if the restrictions on its type parameter are different). Is there a workaround for this? I think we may have talked about this recently, but I can't recall.
- What is the difference between your second snippet and simply calling `do_send_async`?

Comment (thomaseizinger) [reply daemon/src/send_async_safe.rs]: > I cannot implement the same trait for the same generic type twice (even if the restrictions on its type parameter are different). Is there a workaround for this? I think we may have talked about this recently, but I can't recall.

Not even if the trait is parameterized with different things? Note how the 2nd type parameter in `SendAsyncSafe` is different between the two impl blocks.

> What is the difference between your second snippet and simply calling do_send_async?

The 2nd impl will log an error if the async invocation fails whereas `do_send_async` will just drop the result, regardless of what it is.

Comment (luckysori) [reply daemon/src/send_async_safe.rs]: > Not even if the trait is parameterized with different things? Note how the 2nd type parameter in SendAsyncSafe is different between the two impl blocks.

I was doing something wrong and your tip fixed it!

> The 2nd impl will log an error if the async invocation fails whereas do_send_async will just drop the result, regardless of what it is.

Of course, that makes sense. Thanks!

File: daemon/src/maker_cfd.rs [hunk 5]
```diff
@@ -301,6 +302,11 @@ where
             None => None,
         };
 
+        // TODO: How do we get rid of this one? The handler on the
+        // other side is fallible, yet we know that it cannot fail
+        // because we know we just added the connection to this new
+        // taker.
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd.rs]: This is resolved now :)

File: daemon/src/send_async_safe.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            return Err(xtra::Disconnected);
+        }
+
+        let send_fut = self.send(msg);
+
+        #[allow(clippy::disallowed_method)]
+        tokio::spawn(async {
+            let e = match send_fut.await {
+                Ok(Err(e)) => format!("{:#}", e),
+                Err(e) => format!("{:#}", e),
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs:48]: This case is extremely unlikely I guess?

File: daemon/src/send_async_safe.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                Ok(Ok(())) => return,
+            };
+
+            tracing::warn!("Async message invocation failed: {:#}", e)
```
Comment (thomaseizinger) [inline daemon/src/send_async_safe.rs:52]: Looking forward to the `Name` trait adding some context here. (Although I think using `Display` wouldn't be wrong either: https://doc.rust-lang.org/std/fmt/index.html#fmtdisplay-vs-fmtdebug)

**Review by thomaseizinger — APPROVED** [review]
Nice work!

Eventually, it would be nice to migrate all these xtra utilities into an `xtras` crate :drum:

File: daemon/src/send_async_safe.rs (hunk 6 above)
Comment (luckysori) [reply daemon/src/send_async_safe.rs:48]: You're right, it's almost impossible with the check at the top of this method. Worth expecting the first error then?

File: daemon/src/send_async_safe.rs (hunk 7 above)
Comment (luckysori) [reply daemon/src/send_async_safe.rs:52]: I can see `Display` working for `Actor`s, since you will almost always just want the name. For some reason I'd expect `Message` impls of `Display` to also show the contents of the message though.

Comment (thomaseizinger) [reply daemon/src/send_async_safe.rs:52]: For messages, I think I could be convinced either way :D

**Review by thomaseizinger — APPROVED** [review]
